### ISSUE TITLE
demo finds 0 matches

### ISSUE THREAD

**Issue by alrikai** [IntelVCL/dcflow#3]
Running the provided demo on both the KITTI and Sintel frames yields 0 matches; based on some preliminary investigation, it looks like the features computed by the provided convolutional net are identical for the two frames to compute the flow between (despite the frames themselves being different). This happens when running both the KITTI and Sintel frames.  

I am not sure if this is due to some difference or error in my project setup -- are there specific versions of caffe or MATLAB that have to be used? Or for that matter, any specific versions of other dependencies (e.g. OpenCL)? My caffe and matcaffe build compiles and links without errors, and the dcflow code also builds and runs without errors, it just doesn't produce correct results.  

TLDR are there any specific setup requirements, or any reasons for why the demo would not find any matches on the provided data?

**Comment by jia-xu** [comment]
Which GPU,  OpenCL version are you using? We require ~4GB gpu memory to process the cost volume.

Tested with Ubuntu 14.04, 16.04, on NVidia TITAN X, and 1080.

**Comment by alrikai** [comment]
I am using a GTX 1080 and running openSUSE 42.2. I was under the impression the project didn't use opencv based on the CMake files, but I have version 3.1 installed on my system.  

In case it's useful, I am using CUDA 8.0 (which I believe corresponds to OpenCL 1.2), and MATLAB 2017b. Thanks for your help.

**Comment by jia-xu** [comment]
I meant OpenCL version ;)

Can you provide the output from matlab, when you run demo.m? The last few lines should help diagnosing.

Also, can you run nvidia-smi while the demo code is running, and check whether DCFLOW is actually using your GPU?

**Comment by alrikai** [comment]
The text output from running the demo is:
145/341/80
Matching: 289 ms
SGM (fw+bw): 2317 ms

Is this what you meant by the output? Or do you mean the workspace variables, or the caffe logging? I can provide those as well if they would be useful.

Polling nvidia-smi during the demo execution did show GPU usage, i.e. the utilization spiked to 100% and the memory usage went from 973MiB idle to 6356MiB, so I believe it is indeed doing something (just not what we want it to)

Here's a snippet of the nvidia-smi output during the execution showing that MATLAB is indeed the process using the GPU memory 

Fri Nov  3 15:10:04 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.66                 Driver Version: 375.66                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  Off  | 0000:03:00.0     Off |                  N/A |
| 23%   40C    P2   161W / 250W |   6356MiB / 11172MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 0000:04:00.0      On |                  N/A |
| 26%   45C    P0    62W / 250W |    422MiB / 11169MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      8143    C   python3                                        149MiB |
|    0     26303    C   /home/alrik/MATLAB/R2017b/bin/glnxa64/MATLAB  5608MiB |

**Comment by jia-xu** [comment]
Your output looks fine.

Can you visualize the forward and backward flow after line 31, in dcflow.m. And see if there is any structure info?

  forward = single(permute(forward, [2, 3, 1]));   backward = single(permute(backward, [2, 3, 1]));

**Comment by alrikai** [comment]
There isn't much information in them -- forward is a 145 x 341 x 2 array of all -4, and backward is all -1, eg. unique(forward) gives -4, and unique(backward) gives -1. 

Looking back through the results, the problem looks to be from the feature extraction step, (line 21) as the output feature maps are identical along each channel, e.g. unique(feat_1(:,:,1)) is -0.0832, unique(feat_1(:,:,2)) is -0.0391, ...., unique(feat_1(:,:,64)) is -0.0046. In addition, feat_1 is identical to feat_2. I would expect the features to vary across channels and vary between images. This leads me to believe there is something amiss in the caffe model, though I'm not sure if it's on my end or an issue with the model as provided.

**Comment by jia-xu** [comment]
Thanks for the update. I have tested this code thousands of times, and I was able to reproduce the result.

Can you test the  following version of caffe, which we used for both training and testing?
https://github.com/Wangyida/caffe/tree/cnn_triplet

**Comment by alrikai** [comment]
Sorry it's taken me so long to get back to this -- turns out the caffe version was the reason. The demo works quite nicely using your caffe version, and hooking up the matches file with the epicflow visualization code shows reasonable results. Thanks for the help.

**Comment by jia-xu** [comment]
Good to hear ;)
